Adding tangent and bitangent support to glTFFileLoader and Babylon.js

dewadswo opened the pull request:
- Added tangent support to glTFFileLoader
- Added bitangent calc to glTFFileLoader
- Added tangent/bitangent support to pbrMaterials and standardMaterials
- Added transformation for Vector4 in babylon.math.ts
- Added tangent and bitangent support to babylon.mesh.vertexData and babylon.vertexBuffer
- Created bumpVertex and bumpVertexDelaration to shaderIncludes
- Added bumpVertex* support to pbr and standard shaders

dewadswo committed:
Adding tangent and bitangent support to glTFFileLoader and Babylon.js

bghgary commented on loaders/src/glTF/babylon.glTFFileLoader.ts:
```diff
@@ -592,6 +596,24 @@ module BABYLON {
             }
 
+            //bitangent aka binormal
+            if (tempVertexData.tangents !== undefined && tempVertexData.tangents !== null) {
```
this is typically just "if (tempVertexData.tangents)"

dewadswo replied:
aah thanks :) still getting used to typescript

bghgary commented on src/Materials/babylon.pbrMaterial.ts:
```diff
@@ -946,6 +948,12 @@
+                    if (mesh.isVerticesDataPresent(VertexBuffer.TangentKind)) {
+                        this._defines.TANGENT = true;
+                    }
+                    if (mesh.isVerticesDataPresent(VertexBuffer.BiTangentKind)) {
```
nit: Bitangent is one word, so I would expect T to be lowercase

dewadswo replied:
it's my funny camel casing thanks :)

bghgary commented on src/Mesh/babylon.mesh.vertexData.ts:
```diff
@@ -2095,7 +2165,7 @@
             }
 
             // facetPartitioning reinit if needed
-            if (computeFacetPartitioning) {  
```
as a general rule, please don't fix spaces in the same change as other changes. it makes merging a lot more difficult. for this code base, I would recommend not fixing the spaces since there are a lot of contributors.

bghgary replied:
it's also harder to review

dewadswo replied:
yaw sorry.. still getting used to VSCode, I did a format .. fixed in the coming update :)

bghgary replied:
I'm still seeing a few space changes. Are you seeing them?

dewadswo replied:
Just check against a clean 2.0 branch. Should clean now

bghgary commented on src/Mesh/babylon.mesh.vertexData.ts:
```diff
@@ -259,6 +282,35 @@
+                    this.bitangents[index + 1] = transformed.y;
+                    this.bitangents[index + 2] = transformed.z;
+                }
+            }
```
I would expect this to calculated in the shader if bitangents are not provided.

bghgary commented on loaders/src/glTF/babylon.glTFFileLoader.ts:
```diff
@@ -592,6 +596,24 @@ module BABYLON {
+                    var n = Vector3.FromArray(tempVertexData.normals, i);
+                    var t = Vector4.FromArray(tempVertexData.tangents, tindex);
+                    //bitangent cross product of normal and tangent
+                    var bitangent = Vector3.Cross(n, t.toVector3().scale((t.w < 0) ? -1 : 1)).asArray();
```
This really should happen in the shader. I realize you are trying to support TBN in the shader, but I think maybe you should support 3 modes in the shader (normal only, normal + tangent4, normal + bitangent + tangent). Feel free to not support normal + bitangent  + tangent just yet since we don't need it.

bghgary replied:
Did you mean to delete the code here for calculating and setting the bitangents?

dewadswo replied:
deleted

dewadswo committed:
- Various naming and spacing fixes.
- added support in bumpVertex.fx for T+B+N and T+N. If only N then falls
back to just N in fragment shader

bghgary commented on src/Shaders/ShadersInclude/bumpVertex.fx:
```diff
@@ -0,0 +1,8 @@
+	#if defined(TANGENT) && defined(BITANGENT) && defined(NORMAL) 
+		vTBN = mat3(tangent.xyz, bitangent, normal);
+	#elif defined(TANGENT) && defined(NORMAL) 
+		vec3 bitangent = cross(normal, tangent.xyz) * ((tangent.w < 0.0) ? -1.0 : 1.0);
```
You can use the sign function instead.

dewadswo replied:
nice! I never thought about that 👍

bghgary commented on src/Mesh/babylon.vertexBuffer.ts:
```diff
@@ -25,6 +26,7 @@
                     case VertexBuffer.UV6Kind:
                         stride = 2;
                         break;
+                    case VertexBuffer.TangentKind:
```
Do you expect a vec4 for tangents even when using the full TBN?

bghgary replied:
Perhaps you should just remove full TBN support for now to avoid this problem.

dewadswo replied:
done. It's much cleaner now

dewadswo committed:
- Remove bitangent in vertexBuffer/Data, just handle bitangent in shader.
- use sign for handyness of tangent
